go/libraries/doltcore/sqle/index_driver.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2020 Liquidata, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sqle
+
+import (
+	"github.com/src-d/go-mysql-server/sql"
+
+	"github.com/liquidata-inc/dolt/go/libraries/doltcore/schema"
+)
+
+type DoltIndexDriver struct {
+	dbs map[string]Database
+}
+
+var _ sql.IndexDriver = (*DoltIndexDriver)(nil)
+
+func NewDoltIndexDriver(dbs ...Database) *DoltIndexDriver {
+	nameToDB := make(map[string]Database)
+	for _, db := range dbs {
+		nameToDB[db.Name()] = db
+	}
+
+	return &DoltIndexDriver{nameToDB}
+}
+
+func (*DoltIndexDriver) Create(string, string, string, []sql.Expression, map[string]string) (sql.Index, error) {
+	panic("index driver create path not supported")
+}
+
+func (i *DoltIndexDriver) Delete(sql.Index, sql.PartitionIter) error {
+	panic("index driver delete path not supported")
+}
+
+func (*DoltIndexDriver) ID() string {
+	return "doltDbIndexDriver"
+}
+
+func (driver *DoltIndexDriver) LoadAll(ctx *sql.Context, db, table string) ([]sql.Index, error) {
+	database, ok := driver.dbs[db]
+	if !ok {
+		panic("Unexpected db: " + db)
+	}
+
+	root, err := database.GetRoot(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	tbl, ok, err := root.GetTable(ctx, table)
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		return nil, nil
+	}
+
+	sch, err := tbl.GetSchema(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var sqlIndexes []sql.Index
+	for i := 0; i < sch.GetPKCols().Size(); i++ {
```

Comment:
This partial matching of primary keys is a separable concern. It will be good to have, but let's land your new code path for secondary indexes and then iterate. In the meantime just revert the existing working PK index lookup back to their old boring format and add a TODO.

Reply:
Partial matching of primary keys and indexes are related due to their implementation. As you know, indexes are basically tables with every column being a primary key, so they necessitate the ability to partially match a primary key.

I can't really revert the code without deleting the current, working index usage. For the most part, adding true partial key support will just be adding/changing the `IndexLookupKeyIterator`, with everything else remaining fairly as-is.

Reply:
You're confusing the implementation details with the features. The implementation for secondary indexes involves partial primary key matching at the storage layer. As a feature, we can support matching *all* of the columns in an index, or *some subset* of the columns in an index. What I'm saying is that we should start by supporting just matching *all* the columns in an index, and add support for partial matches later. When we do that, it might make more sense to implement at least partially in the go-mysql-server layer, rather than declaring each partial index to the framework separately, as you're doing now. I'm not sure yet, let's decide later. Start small and iterate. You have enough complexity and testing burden here for 10 PRs already.